File: others/100-200/132. 分割回文串 II.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from typing import *
from queue import *

MOD = int(1e9 + 7)
DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]
INF = int(1e9)


class Solution:
    def minCut(self, s: str) -> int:
        n = len(s)
        is_p = [[False] * n for _ in range(n)]
        for i in range(n - 1, -1, -1):
            for j in range(i, n):
                if s[i] == s[j] and (j - i < 2 or is_p[i + 1][j - 1]):
                    is_p[i][j] = True

        dp = [INF] * (n + 1)
        dp[0] = 0
        for i in range(1, n + 1):
            for j in range(i):
                if is_p[j][i - 1]:
                    dp[i] = min(dp[i], dp[j] + 1)
        return dp[n] - 1


class Solution:
    def minCut(self, s: str) -> int:
        n = len(s)
        # 预处理回文子串
        is_pal = [[False] * n for _ in range(n)]
        for i in range(n - 1, -1, -1):
            for j in range(i, n):
                if s[i] == s[j] and (j - i < 2 or is_pal[i + 1][j - 1]):
                    is_pal[i][j] = True

        # dp[i]: s[:i] 的最小切割数
        dp = [float("inf")] * (n + 1)
        dp[0] = -1  # 空串前切一次，回头 +1 抵消

        for i in range(1, n + 1):
            for j in range(i):
                if is_pal[j][i - 1]:
                    dp[i] = min(dp[i], dp[j] + 1)
        return dp[n]

File: others/100-200/132. 分割回文串 II.cpp
//
// Created by 123 on 25-6-29.
//
#include "bits/stdc++.h"
using namespace std;
using ll = long long;
constexpr int MOD = 1'000'000'007;
constexpr int DIR[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

class Solution {
public:
	int minCut(string s) {
		int n = s.size();
		vector<vector<bool> > is_p(n, vector<bool>(n));
		for (int i = 0; i < n; i++) {
			is_p[i][i] = true;
			for (int j = 0; j < i; j++) {
				if (s[j] == s[i] && (is_p[j + 1][i - 1] || j + 1 == i)) {
					is_p[j][i] = true;
				}
			}
		}

		vector<int> dp(n + 1,INT_MAX);
		dp[0] = -1;
		for (int i = 1; i <= n; i++) {
			for (int j = 1; j <= i; j++) {
				if (is_p[j- 1][i - 1]) {
					dp[i] = min(dp[i], dp[j - 1] + 1);
				}
			}
		}

		return dp[n];
	}
};
